### PR TITLE
[otlp] Remove the Google.Protobuf / Grpc packages, and replace the logs and metrics with the new implementation

### LINKED ISSUE #5730: Custom serializer for OTLP exporter
### Package

OpenTelemetry.Exporter.OpenTelemetryProtocol

### Is your feature request related to a problem?

Relates to https://github.com/open-telemetry/opentelemetry-dotnet/issues/4395.

Depending on Google.Protobuf has multiple issues associated with it. #4395 notes the impact on auto-instrumentation project. In addition to it, it contributes to increased package size and overhead maintenance of reacting to CVEs when they arise. 

Also, the performance overhead of using a general-purpose serializer is suboptimal.

Opening this issue to track the work to remove the dependency on Google.Protobuf package and implement our own custom serialize based on the [OTLP protocol specification](https://github.com/open-telemetry/opentelemetry-proto)

### What is the expected behavior?

A high performance serializer specifically designed to comply with OTLP specification requirements.

### Which alternative solutions or features have you considered?

NA

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rajkumar-rangaraj**
Fixes #5730
Design discussion issue #

## Changes

Please provide a brief description of the changes here.

#### OpenTelemetry.Exporter.OpenTelemetryProtocol

* Replaced the current logs and metrics implementation with the new one. Deleted all existing implementations and renamed the new implementation to the existing names.
* Remove the following package references from the OTLP Exporter project
   * Google.Protobuf
   * Grpc
   * Grpc.Net.Client
   * Grpc.Tools
* Deleted all proto files from the project

#### Tests / Benchmarks
* Updated both tests and benchmarks project to include proto files and the following packages `Google.Protobuf`, `Grpc`, `Grpc.Tools` and `Grpc.Net.Client`.
* Updated the tests.

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpExportClient.cs [hunk 1]
```diff
@@ -28,13 +28,24 @@ static ProtobufOtlpExportClient()
...
-    protected ProtobufOtlpExportClient(OtlpExporterOptions options, HttpClient httpClient, string signalPath)
+    protected OtlpExportClient(OtlpExporterOptions options, HttpClient httpClient, string signalPath)
     {
         Guard.ThrowIfNull(options);
         Guard.ThrowIfNull(httpClient);
         Guard.ThrowIfNull(signalPath);
 
-        Uri exporterEndpoint = options.Endpoint.AppendPathIfNotPresent(signalPath);
+        Uri exporterEndpoint;
+        if (options.Protocol == OtlpExportProtocol.Grpc)
```
Comment (rajkumar-rangaraj) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpExportClient.cs:38]: Added a logical condition here to honor the existing behavior.

File: test/Benchmarks/Benchmarks.csproj [hunk 2]
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFrameworks>$(TargetFrameworksForTests)</TargetFrameworks>
+    <NoWarn>$(NoWarn);SA1518;SA1210</NoWarn>
```
Comment (rajkumar-rangaraj) [inline test/Benchmarks/Benchmarks.csproj:6]: @CodeBlanch Please validate if this is a correct way to suppress the proto file message.

**Comment by TimothyMothra** [comment]
It looks like the proto files are copied into two directories, unit test & benchmarks.
Can these be stored in a common directory?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpExportClient.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpExportClient.cs:38]: Curious how it go by tests before?

File: .editorconfig [hunk 3]
```diff
@@ -156,7 +156,7 @@ dotnet_diagnostic.IDE0005.severity = warning
 # RS0041: Public members should not use oblivious types
 dotnet_diagnostic.RS0041.severity = suggestion
 
-[obj/**.cs]
+[**/obj/**.cs]
```
Comment (CodeBlanch) [inline .editorconfig:159]: @rajkumar-rangaraj This cleans up those proto warnings you had to suppress in project files.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OpenTelemetry.Exporter.OpenTelemetryProtocol.csproj [hunk 4]
```diff
@@ -23,13 +23,6 @@
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry\OpenTelemetry.csproj" />
   </ItemGroup>
 
-  <ItemGroup>
-    <PackageReference Include="Grpc.Net.Client" Condition="'$(TargetFramework)' != 'netstandard2.0' AND '$(TargetFramework)' != '$(NetFrameworkMinimumSupportedVersion)'" />
-    <PackageReference Include="Grpc" Condition="'$(TargetFramework)' == 'netstandard2.0' OR '$(TargetFramework)' == '$(NetFrameworkMinimumSupportedVersion)'" />
-    <PackageReference Include="Google.Protobuf" />
-    <PackageReference Include="Grpc.Tools" PrivateAssets="All" />
-  </ItemGroup>
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OpenTelemetry.Exporter.OpenTelemetryProtocol.csproj:31]: Dropping these dependencies is probably worthy of a CHANGELOG entry? What you think @alanwest?

Comment (rajkumar-rangaraj) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OpenTelemetry.Exporter.OpenTelemetryProtocol.csproj:31]: This definitely needs to be a part of the changelog. There are few things we need to cover in the changelog, planning to cover as follow-up.

1. Removal of this packages
2. Breaking change on PeerServiceResolver, in one the earlier PR got feedback from you and Alan and we decided to remove this look up as we aren't aware why this logic was present https://github.com/open-telemetry/opentelemetry-dotnet/blob/84e6afbebae67f4e9b498f1702edfe4ceccf34ec/src/Shared/PeerServiceResolver.cs
3. Call-out on the internal serialization change.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpExportClient.cs (hunk 1 above)
Comment (rajkumar-rangaraj) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/OtlpExportClient.cs:38]: The actual issue was with the HTTP implementation. In the gRPC case, we append the path in all scenarios. However, for HTTP, the path is appended only if it is provided by default or through environment variables. If an endpoint is explicitly set, we do not append the path in the HTTP case, but this behavior was not being followed in new implementation.

During my initial implementation, I mistakenly assumed this was a bug and removed the check. However, there was an existing test case that validated this behavior.

```
exporterEndpoint = options.AppendSignalPathToEndpoint
                ? options.Endpoint.AppendPathIfNotPresent(signalPath)
                : options.Endpoint;
```

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Pull request merged**